tests/test_s3urlcache.py

```diff
@@ -0,0 +1,196 @@
+
+    def test_store_in_cache(self):
+        """The URL contents are stored in S3 and the contents returned, when requested url is not found in cache."""
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        url_key = S3UrlCache._url_to_key(url)
+
+        self._delete_cached_urls()
+        with self.assertLogs(logger, "INFO") as log_monitor:
+            url_content = self.cache.resolve(url)
+
+        original_data = randomdata[:KiB]
+        self.assertEqual(len(url_content), KiB)
+        self.assertEqual(url_content, original_data)
+        self.assertTrue(log_monitor.output[0].endswith(f"{url} not found in cache. Adding it to "
+                                                       f"{self.test_bucket} with key {url_key}."))
+
+    def test_retrieve_from_cache(self):
+        """Stored URL contents is retrieved from S3, when a cached url is requested"""
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        with self.assertLogs(logger, 'INFO') as log_monitor:
+            url_content = self.cache.resolve(url)
+            cached_content = self.cache.resolve(url)
+        self.assertEqual(len(log_monitor.output), 1)
+        self.assertEqual(url_content, cached_content)
+
+    def test_bad_url(self):
+        bad_urls = ['', '//', 'http://?']
+        self.urls_to_cleanup.update(bad_urls)
+
+        for url in bad_urls:
+            with self.subTest(bad_url=url):
+                with self.assertRaises(requests.RequestException):
+                    self.cache.resolve(url)
+
+    def test_url_SizeLimitError(self):
+        """Exception returned when URL content size is greater than max_size."""
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        self.cache.max_size = 1
+        with self.assertRaises(SizeLimitError) as ex:
+            self.cache.resolve(url)
+        self.assertEqual(ex.exception.args[0], f"{url} not cached. The URL's contents have exceeded "
+                                               f"{self.cache.max_size} bytes.")
+
+    def test_content_size_0(self):
+        url = f"{HTTPInfo.url}/0"
+        self.urls_to_cleanup.add(url)
+        url_content = self.cache.resolve(url)
+        self.assertEqual(len(url_content), 0)
+
+    def test_chunked_content(self):
+        size = KiB * 10
+        url = f"{HTTPInfo.url}/{size}"
+        self.urls_to_cleanup.add(url)
+        self.cache.max_size = 1 * MB
+        self.cache.chunk_size = 1 * KiB
+        url_content = self.cache.resolve(url)
+        original_data = randomdata[:size]
+        self.assertEqual(url_content, original_data)
+
+    def test_stored_url_metadata(self):
+        url = f"{HTTPInfo.url}/{KiB}"
+        self.urls_to_cleanup.add(url)
+        url_key = S3UrlCache._url_to_key(url)
+        self.cache.resolve(url)
+
+        with self.subTest("check dss_cached_url"):
+            metadata = self.cache._reverse_key_lookup(url_key)
```

Comment:
s/metadata/cached_url

The term metadata leaks an implementation detail.

Reply:
fixed